File: crates/mithras-crypto/src/utxo.rs
use curve25519_dalek::Scalar;

pub const SECRET_SIZE: usize = 136;
use ed25519_dalek::VerifyingKey as Ed25519PublicKey;
use hpke_rs::HpkePublicKey;

use crate::{
    MithrasError,
    address::MithrasAddr,
    discovery::{compute_discovery_secret_sender, compute_discovery_tag},
    hpke::{HpkeEnvelope, SupportedHpkeSuite},
    keypairs::{DiscoveryKeypair, derive_tweak_scalar, derive_tweaked_pubkey},
};

#[derive(Clone, Debug, PartialEq)]
pub struct UtxoSecrets {
    pub spending_secret: [u8; 32],
    pub nullifier_secret: [u8; 32],
    pub amount: u64,
    pub tweak_scalar: Scalar,
    pub tweaked_pubkey: Ed25519PublicKey,
}

impl UtxoSecrets {
    pub fn from_hpke_envelope(
        hpke_envelope: HpkeEnvelope,
        discovery_keypair: DiscoveryKeypair,
        txn_metadata: &crate::hpke::TransactionMetadata,
    ) -> Result<Self, MithrasError> {
        let hpke = SupportedHpkeSuite::Base25519Sha512ChaCha20Poly1305.suite();

        let hpke_recipient_private =
            hpke_rs::HpkePrivateKey::new(discovery_keypair.private_key().to_bytes().to_vec());

        let mut recv_ctx = hpke
            .setup_receiver(
                &hpke_envelope.encapsulated_key,
                &hpke_recipient_private,
                &txn_metadata.info(),
                None,
                None,
                None,
            )
            .map_err(|e| MithrasError::HpkeOperation { msg: e.to_string() })?;

        let pt = recv_ctx
            .open(&txn_metadata.aad(), &hpke_envelope.ciphertext)
            .map_err(|e| MithrasError::HpkeOperation { msg: e.to_string() })?;
        let pt_array: [u8; SECRET_SIZE] =
            pt.try_into().map_err(|_| MithrasError::DataConversion {
                msg: "Invalid plaintext size for UTXO secrets".to_string(),
            })?;
        UtxoSecrets::try_from(pt_array)
    }
}

impl TryFrom<[u8; SECRET_SIZE]> for UtxoSecrets {
    type Error = MithrasError;

    fn try_from(bytes: [u8; SECRET_SIZE]) -> Result<Self, Self::Error> {
        let mut spending_secret = [0u8; 32];
        let mut nullifier_secret = [0u8; 32];

        spending_secret.copy_from_slice(&bytes[0..32]);
        nullifier_secret.copy_from_slice(&bytes[32..64]);
        let amount = u64::from_be_bytes(bytes[64..72].try_into().map_err(|_| {
            MithrasError::DataConversion {
                msg: "Failed to convert amount bytes".to_string(),
            }
        })?);
        let tweak_scalar =
            Scalar::from_bytes_mod_order(bytes[72..104].try_into().map_err(|_| {
                MithrasError::DataConversion {
                    msg: "Failed to convert tweak scalar bytes".to_string(),
                }
            })?);
        let tweak_pubkey =
            Ed25519PublicKey::from_bytes(&bytes[104..136].try_into().map_err(|_| {
                MithrasError::DataConversion {
                    msg: "Failed to convert tweaked pubkey bytes".to_string(),
                }
            })?)
            .map_err(|e| MithrasError::Ed25519KeyParsing { msg: e.to_string() })?;

        Ok(Self {
            spending_secret,
            nullifier_secret,
            amount,
            tweak_scalar,
            tweaked_pubkey: tweak_pubkey,
        })
    }
}

impl From<UtxoSecrets> for [u8; SECRET_SIZE] {
    fn from(secret: UtxoSecrets) -> Self {
        let mut bytes = [0u8; SECRET_SIZE];
        bytes[0..32].copy_from_slice(&secret.spending_secret);
        bytes[32..64].copy_from_slice(&secret.nullifier_secret);
        bytes[64..72].copy_from_slice(&secret.amount.to_be_bytes());
        bytes[72..104].copy_from_slice(&secret.tweak_scalar.to_bytes());
        bytes[104..136].copy_from_slice(&secret.tweaked_pubkey.to_bytes());
        bytes
    }
}

#[derive(Debug, PartialEq)]
pub struct UtxoInputs {
    pub secrets: UtxoSecrets,
    pub hpke_envelope: HpkeEnvelope,
}

impl UtxoInputs {
    pub fn generate(
        txn_metadata: &crate::hpke::TransactionMetadata,
        amount: u64,
        receiver: &MithrasAddr,
    ) -> Result<Self, MithrasError> {
        let mut hpke = SupportedHpkeSuite::Base25519Sha512ChaCha20Poly1305.suite();

        let ephemeral_keypair = DiscoveryKeypair::generate()?;

        let discovery_secret =
            compute_discovery_secret_sender(ephemeral_keypair.private_key(), &receiver.disc_x25519);

        let tweak_scalar = derive_tweak_scalar(&discovery_secret);

        let tweaked_pubkey = derive_tweaked_pubkey(&receiver.spend_ed25519, &tweak_scalar)?;

        let discovery_tag = compute_discovery_tag(
            &discovery_secret,
            &txn_metadata.sender,
            txn_metadata.first_valid,
            txn_metadata.last_valid,
            txn_metadata.lease,
        )?;

        // TODO: ensure secrets are in scalar field
        let mut spending_secret = [0u8; 32];
        getrandom::fill(&mut spending_secret)
            .map_err(|e| MithrasError::RandomGeneration { msg: e.to_string() })?;

        let mut nullifier_secret = [0u8; 32];
        getrandom::fill(&mut nullifier_secret)
            .map_err(|e| MithrasError::RandomGeneration { msg: e.to_string() })?;

        let (encapsulated_key, mut sender_ctx) = hpke
            .setup_sender(
                &HpkePublicKey::new(receiver.disc_x25519.to_bytes().to_vec()),
                &txn_metadata.info(),
                None,
                None,
                None,
            )
            .map_err(|e| MithrasError::HpkeOperation { msg: e.to_string() })?;

        let mithras_secret = UtxoSecrets {
            spending_secret,
            nullifier_secret,
            amount,
            tweak_scalar,
            tweaked_pubkey,
        };

        let secret_bytes: [u8; SECRET_SIZE] = mithras_secret.clone().into();
        let ct = sender_ctx
            .seal(&txn_metadata.aad(), &secret_bytes)
            .map_err(|e| MithrasError::HpkeOperation { msg: e.to_string() })?;

        let hpke_envelope = HpkeEnvelope {
            version: 1,
            suite: SupportedHpkeSuite::Base25519Sha512ChaCha20Poly1305,
            encapsulated_key: encapsulated_key.try_into().map_err(|_| {
                MithrasError::DataConversion {
                    msg: "Invalid encapsulated key size".to_string(),
                }
            })?,
            ciphertext: ct.try_into().map_err(|_| MithrasError::DataConversion {
                msg: "Invalid ciphertext size".to_string(),
            })?,
            discovery_tag,
        };

        Ok(UtxoInputs {
            secrets: mithras_secret,
            hpke_envelope,
        })
    }
}

File: packages/typescript/mithras-crypto/src/utxo.ts
import {
  bytesToNumberBE,
  concatBytes,
  numberToBytesBE,
} from "@noble/curves/utils.js";
import {
  getHpkeSuite,
  HpkeEnvelope,
  SupportedHpkeSuite,
  TransactionMetadata,
} from "./hpke";
import {
  deriveTweakedPubkey,
  deriveTweakScalar,
  DiscoveryKeypair,
} from "./keypairs";
import { computeDiscoverySecretSender, computeDiscoveryTag } from "./discovery";
import { MithrasAddr } from "./address";
import { mimcSum } from "./mimc";

export const SECRET_SIZE: number = 136;

export class UtxoSecrets {
  spendingSecret: Uint8Array;
  nullifierSecret: Uint8Array;
  amount: bigint;
  tweakScalar: bigint;
  tweakedPubkey: Uint8Array;

  constructor(
    spendingSecret: Uint8Array,
    nullifierSecret: Uint8Array,
    amount: bigint,
    tweakScalar: bigint,
    tweakedPubkey: Uint8Array,
  ) {
    this.spendingSecret = spendingSecret;
    this.nullifierSecret = nullifierSecret;
    this.amount = amount;
    this.tweakScalar = tweakScalar;
    this.tweakedPubkey = tweakedPubkey;
  }

  static fromBytes(bytes: Uint8Array): UtxoSecrets {
    if (bytes.length !== SECRET_SIZE) {
      throw new Error(
        `Invalid byte array length for UtxoSecrets: expected ${SECRET_SIZE}, got ${bytes.length}`,
      );
    }

    const spendingSecret = bytes.slice(0, 32);
    const nullifierSecret = bytes.slice(32, 64);
    const amount = bytesToNumberBE(bytes.slice(64, 72));
    const tweakScalar = bytesToNumberBE(bytes.slice(72, 104));
    const tweakedPubkey = bytes.slice(104, 136);

    return new UtxoSecrets(
      spendingSecret,
      nullifierSecret,
      amount,
      tweakScalar,
      tweakedPubkey,
    );
  }

  toBytes(): Uint8Array {
    const bytes = new Uint8Array(SECRET_SIZE);
    bytes.set(this.spendingSecret, 0);
    bytes.set(this.nullifierSecret, 32);
    bytes.set(numberToBytesBE(this.amount, 8), 64);
    bytes.set(numberToBytesBE(this.tweakScalar, 32), 72);
    bytes.set(this.tweakedPubkey, 104);
    return bytes;
  }

  static async fromHpkeEnvelope(
    hpkeEnvelope: HpkeEnvelope,
    discoveryKeypair: DiscoveryKeypair,
    txnMetadata: TransactionMetadata,
  ): Promise<UtxoSecrets> {
    const hpke = getHpkeSuite(hpkeEnvelope.suite);

    const recvCtx = await hpke.createRecipientContext({
      recipientKey: await hpke.kem.deserializePrivateKey(
        discoveryKeypair.privateKey,
      ),
      enc: hpkeEnvelope.encapsulatedKey,
      info: txnMetadata.info(),
    });

    const plaintext = await recvCtx.open(
      hpkeEnvelope.ciphertext,
      txnMetadata.aad(),
    );

    return UtxoSecrets.fromBytes(new Uint8Array(plaintext));
  }

  // fn compute_nullifier(utxo: &UtxoSecrets) -> [u8; 32] {
  //     mimc::mimc_sum(&[compute_commitment(utxo), utxo.nullifier_secret])
  // }
  //
  // fn compute_commitment(utxo: &UtxoSecrets) -> [u8; 32] {
  //     let mut amount_bytes = [0u8; 32];
  //     amount_bytes[24..].copy_from_slice(&utxo.amount.to_be_bytes());
  //
  //     mimc::mimc_sum(&[
  //         utxo.spending_secret,
  //         utxo.nullifier_secret,
  //         amount_bytes,
  //         utxo.tweaked_pubkey.to_bytes(),
  //     ])
  // }

  computeCommitment(): Uint8Array {
    return numberToBytesBE(
      mimcSum([
        bytesToNumberBE(this.spendingSecret),
        bytesToNumberBE(this.nullifierSecret),
        this.amount,
        bytesToNumberBE(this.tweakedPubkey),
      ]),
      32,
    );
  }

  computeNullifier(): Uint8Array {
    const commitment = this.computeCommitment();
    return numberToBytesBE(
      mimcSum([
        bytesToNumberBE(commitment),
        bytesToNumberBE(this.nullifierSecret),
      ]),
      32,
    );
  }
}

export class UtxoInputs {
  secrets: UtxoSecrets;
  hpkeEnvelope: HpkeEnvelope;

  private constructor(secrets: UtxoSecrets, hpkeEnvelope: HpkeEnvelope) {
    this.secrets = secrets;
    this.hpkeEnvelope = hpkeEnvelope;
  }

  static async generate(
    txnMetadata: TransactionMetadata,
    amount: bigint,
    receiver: MithrasAddr,
  ): Promise<UtxoInputs> {
    const hpke = getHpkeSuite(SupportedHpkeSuite.x25519Sha256ChaCha20Poly1305);

    const ephemeralKeypair = DiscoveryKeypair.generate();

    const discoverySecret = computeDiscoverySecretSender(
      ephemeralKeypair.privateKey,
      receiver.discX25519,
    );

    const tweakScalar = deriveTweakScalar(discoverySecret);
    const tweakedPubkey = deriveTweakedPubkey(
      receiver.spendEd25519,
      tweakScalar,
    );

    const discoveryTag = computeDiscoveryTag(
      discoverySecret,
      txnMetadata.sender,
      txnMetadata.firstValid,
      txnMetadata.lastValid,
      txnMetadata.lease,
    );

    const spendingSecret = crypto.getRandomValues(new Uint8Array(32));
    const nullifierSecret = crypto.getRandomValues(new Uint8Array(32));

    const mithrasSecret = new UtxoSecrets(
      spendingSecret,
      nullifierSecret,
      amount,
      tweakScalar,
      tweakedPubkey,
    );

    const senderCtx = await hpke.createSenderContext({
      recipientPublicKey: await hpke.kem.deserializePublicKey(
        receiver.discX25519,
      ),
      info: txnMetadata.info(),
    });

    const ct = await senderCtx.seal(mithrasSecret.toBytes(), txnMetadata.aad());

    const hpkeEnvelope = new HpkeEnvelope(
      1,
      SupportedHpkeSuite.x25519Sha256ChaCha20Poly1305,
      new Uint8Array(senderCtx.enc),
      new Uint8Array(ct),
      discoveryTag,
      ephemeralKeypair.publicKey,
    );

    return new UtxoInputs(mithrasSecret, hpkeEnvelope);
  }
}
